File: src/models/Review.ts
import mongoose from 'mongoose';

const reviewSchema = new mongoose.Schema({
    productId: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
    customerName: { type: String, required: true },
    rating: { type: Number, required: true, min: 1, max: 5 },
    comment: { type: String, default: '' },
    createdAt: { type: Date, default: Date.now }
});

// @database-optimizer: Indexes for product review queries
reviewSchema.index({ productId: 1, createdAt: -1 }); // Product reviews sorted by newest
reviewSchema.index({ productId: 1, rating: -1 }); // Product reviews sorted by rating
reviewSchema.index({ createdAt: -1 }); // Admin review listing

const Review = mongoose.models.Review || mongoose.model('Review', reviewSchema);

export default Review;

File: ecom_app/lib/models/review.dart
class Review {
  final String id;
  final String productId;
  final String productName;
  final String userName;
  final int rating;
  final String comment;
  final DateTime createdAt;

  Review({
    required this.id,
    required this.productId,
    required this.productName,
    required this.userName,
    required this.rating,
    required this.comment,
    required this.createdAt,
  });

  factory Review.fromJson(Map<String, dynamic> json) {
    return Review(
      id: json['_id'] ?? '',
      productId: json['productId']?['_id'] ?? json['productId'] ?? '',
      productName: json['productId']?['name'] ?? 'Unknown Product',
      userName: json['userName'] ?? 'Anonymous',
      rating: json['rating'] ?? 0,
      comment: json['comment'] ?? '',
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }
}
